fortuna/src/keeper.rs

```diff
@@ -0,0 +1,388 @@
+                to:   latest_safe_block,
+            },
+            backlog_contract,
+            chain_eth_config.gas_limit,
+            backlog_chain_state.clone(),
+        )
+        .await;
+        tracing::info!(
+            "Backlog processing for chain: {} completed",
+            &backlog_chain_state.id
+        );
+    });
+
+    let (tx, rx) = mpsc::channel::<BlockRange>(1000);
+
+    let watch_blocks_chain_state = chain_state.clone();
+    spawn(async move {
+        loop {
+            if let Err(e) = watch_blocks(
+                watch_blocks_chain_state.clone(),
+                latest_safe_block,
+                tx.clone(),
+                chain_eth_config.geth_rpc_wss.clone(),
+            )
+            .await
+            {
+                tracing::error!(
+                    "Error in watching blocks for chain: {}, {:?}",
+                    &watch_blocks_chain_state.id,
+                    e
+                );
+                time::sleep(RETRY_INTERVAL).await;
+            }
+        }
+    });
+    spawn(process_new_blocks(
+        chain_state.clone(),
+        rx,
+        Arc::clone(&contract),
+        chain_eth_config.gas_limit,
+    ));
+}
+
+pub async fn process_event(
+    event: RequestedWithCallbackEvent,
+    chain_config: &BlockchainState,
+    contract: &Arc<SignablePythContract>,
+    gas_limit: U256,
+) -> Result<()> {
+    if chain_config.provider_address != event.provider_address {
+        return Ok(());
+    }
+    let provider_revelation = match chain_config.state.reveal(event.sequence_number) {
+        Ok(result) => result,
+        Err(e) => {
+            tracing::error!(
+                "Error while revealing for provider: {} and sequence number: {} with error: {:?}",
+                event.provider_address,
+                event.sequence_number,
+                e
+            );
+            return Ok(());
+        }
+    };
+
+    let gas_estimate_res = chain_config
+        .contract
+        .estimate_reveal_with_callback_gas(
+            event.provider_address,
+            event.sequence_number,
+            event.user_random_number,
+            provider_revelation,
+        )
+        .await;
+
+    match gas_estimate_res {
+        Ok(gas_estimate_option) => match gas_estimate_option {
+            Some(gas_estimate) => {
+                if gas_estimate > gas_limit {
+                    tracing::error!(
+                        "Gas estimate for reveal with callback is higher than the gas limit for chain: {}",
+                        &chain_config.id
+                    );
+                    return Ok(());
+                }
+
+                let res = contract
+                    .reveal_with_callback(
+                        event.provider_address,
+                        event.sequence_number,
+                        event.user_random_number,
+                        provider_revelation,
+                    )
+                    .gas(gas_estimate)
```

Comment:
does this estimate need to be padded a bit to ensure the TX lands? (I really don't know the answer to this)